model.py

```diff
@@ -1,103 +1,1394 @@
+
+        # Add the parameters to the self.XXX.
+        for par in parameter_objects:
+            setattr(self, par.name, par)
+
+        # Handle the family
+        if children is None:
+            children = []
+        self.children = children
+        self.parent = parent
+
+        # Verify my parameters get put in the right places
+        self.__check_par_assignments__()
+
+    # Tree handling methods
+    def search_par(self, label, name):
+        '''Check if I have <name> in my attributes. If I don't, check my children.
+        If they don't, return None'''
+        # If I'm the desired node, get my parameter
+        if self.label == label:
+            return getattr(self, name)
+        # Otherwise, check my children.
+        else:
+            for child in self.children:
+                val = child.search_par(label, name)
+                if val is not None:
+                    return val
+            return None
+
+    def search_Node(self, class_type, label):
+        '''Search for a node below me of class_type, with the label requested.
+        '''
+        if self.name == "{}_{}".format(class_type, label):
+            return self
+        else:
+            for child in self.children:
+                val = child.search_Node(class_type, label)
+                if val is not None:
+                    return val
+                else:
+                    pass
+            return None
+
+    def search_node_type(self, class_type, nodes=None):
+        '''Construct a set of all the nodes of a given type below me'''
+
+        if nodes is None:
+            nodes = set()
+
+        for child in self.children:
+            child_nodes = child.search_node_type(class_type, nodes)
+            nodes.union(child_nodes)
+
+        if str(self.__class__.__name__) == class_type:
+            nodes.add(self)
+
+        return nodes
+
+    def add_child(self, children):
+        # This check allows XXX.add_child(Param) to be valid
+        if not isinstance(children, list):
+            children = [children]
+
+        self.children.extend(children)
+
+    # Tree evaluation methods
+    def evaluate_model(self, *args, **kwargs):
+        '''Overwrite this method for the nodes capable of evaluating the model.
+        Called by ln_like. Example would be having this call a chisq function
+        and return the output.'''
+
+        if self.is_leaf:
+            msg = "Leaf {} of the model do not have an evaluation function!"
+            msg.format(self.name)
+            raise NotImplementedError(msg)
+        else:
+            raise NotImplementedError
+
+    def ln_like(self, *args, **kwargs):
+        '''Calculate the log likelihood, via chi squared.
+        Extra arguments can be supplied to the model evaluation
+        via args and kwargs.'''
+        ln_like = 0.0
+
+        try:
+            #Execute the model evaluation with a deadline.
+            ln_like += self.evaluate_model(*args, **kwargs)
```

Comment:
The logic at the moment (apologies if this isn't obvious) is:
1. `ln_like` is always present at every node. By default, calculates the sum of the `evaluate_model` functions at and below this node. Stops searching a branch when it successfully calls `evaluate_model`.
2. `evaluate_model` (you're right, the `_model` is redundant) does the grunt work of `ln_like`, and returns the `log-likelihood` of <b>this</b> node. It's implimented on the `Eclipse` and `GPLCModel` classes, which both do this, one with `-0.5 * chisq` and the other with `gp.log_likelihood`
3. `chisq` returns a chisq, but isn't present by default.


Always evaluating leaf nodes made the GP a bit complex. The hyperparameters are obviously shared between eclipses, so they all need to be children of a GP node (I suppose they could all point to the same Param objects for the GP, though that's a fairly abusive solution...). The obvious way to do that it to make the `LCModel` either a `GPLCModel` variant, or a child of a GP node. The trouble then is that you need <i>both</i> a GP node, and a `GPEclipse` variant to handle the `evaluate_model()` call this way, since the Eclipse is the leaf, not the GP node.

By falling down the tree until a valid `evaluate` call, you'd only need the `GPLCModel` since it can stop the recursion pretty much immediately, and can calculate the GP likelihood on each eclipse. I figured this would be more flexible.


Does that make sense? I'm worried about implimenting something that isn't obvious...

Reply:
There are several related issues here that are causing problems:

1) You are mixing and matching modes of operation in which the root node is responsible for evaluating the whole model, and ones in which the leaf nodes do that. This makes it extremely hard to follow the logic of the code. **Don't do this**; pick one and stick to it

2) log_likelihood should only be computed by functions called ```ln_like```. ```evaluate``` should calculate the actual result of the model! I.e ```evaluate``` has to return a lightcurve. Anything where routines do things you don't expect based on their name is confusing.

3) I don't see why **all** of these aggregate functions (ln_like, chisq, etc) can't be implemented on ```Model```, so that any subclass/node which doesn't override the default just sums the value returned by it's children. At the leaf nodes you over-ride these functions to return an actual value and it all works like magic.

You don't need two functions (as you have now) to handle this logic, do you?

4)  Under the model above, GPs are pretty easy to do, no? Standard models have the hierarchy

LCModel -- Band -- SimpleEclipse

or 

LCModel -- Band -- ComplexEclipse

whereas GP-using models have the heirarchy

GPLCModel -- Band -- SimpleGPEclipse

or 

GPLCModel -- Band -- ComplexGPEclipse

The GPLCModel is super simple - it just needs extra params in ```node_param_names``` and no extra logic. The GP eclipse variants have a different ```ln_like```....

Reply:
Would it be more obvious if I rename `evaluate_model` to `self.__ln_like__()` function?

Reply:
It would be more logical, but it doesn't solve the issues with you mixing a top-down and bottom-up approach. Nor does it change the fact that you are using two functions when you only need one. 

Is there a problem with my suggestion that I'm missing? Why not just do it like that?

Reply:
Ah sorry, I'd not refreshed the page and seen your other comment, only the first one. I think I'm with it now, I get the confusion. I'll change the names to the more sensible versions, I think I'd talked myself around into a more confusing naming scheme. 

I'll do that version of the GP, too. I can see how having two conflicting methods is confusing, and it's not like it'll make the code slower or anything.

Reply:
The ln_like thing is, I think, sorted. Please check we're actually on the same page!

I only just saw the time, but I'll finish the GP changes tomorrow. I need to sit down with a pen and paper and sketch out the logic rather than diving in.

Reply:
yes - that's exactly what I meant with ```ln_like``` - can do the same with chisq of course as well!